### PR TITLE
feat(virtctl,create): Add flags to generate cloud-init config 

### PR DESCRIPTION
**Pull request by 0xFelix**
### What this PR does

This adds flags to virtctl create vm which allow creating a basic
cloud-init config so the created VMs become more useful.

The added flags allow to:

- Set the username
- Read the password for the user from a file
- Add one or more SSH authorized keys for the user
- Enable qemu-guest-agent to write to the authorized_keys file of the
  user (SELinux bool)

Furthermore, this addresses a TODO by extending the support cloud-init
data source types, so both noCloud and configDrive types are supported
now.

Before this PR:

cloud-init config cannot be generated with virtctl create vm

After this PR:

Basic cloud-init config can be generated with virtctl create vm

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
virtctl: Allow creating a basic cloud-init config with virtctl create vm
```

### CODE REVIEW

File: pkg/virtctl/create/vm/vm.go [hunk 1]
```diff
@@ -318,8 +229,21 @@ func NewCommand(clientConfig clientcmd.ClientConfig) *cobra.Command {
...
 	))
 
+	cmd.Flags().StringVar(&c.user, UserFlag, c.user, "Specify the user in the cloud-init user data that is added to the VM.")
+	cmd.Flags().BoolVar(&c.genPassword, GenPasswordFlag, c.genPassword, "Specify if a password should be generated in the cloud-init user data that is added to the VM.")
+	cmd.Flags().StringVar(&c.passwordFile, PasswordFileFlag, c.passwordFile, "Specify a file to read the password from for the cloud-init user data that is added to the VM.")
+	cmd.Flags().StringSliceVar(&c.sshKeys, SSHKeyFlag, c.sshKeys, "Specify one or more SSH authorized keys in the cloud-init user data that is added to the VM.")
+	cmd.Flags().BoolVar(&c.gaManageSSH, GAManageSSHFlag, c.gaManageSSH, "Specify if the qemu-guest-agent should be able to manage SSH in the cloud-init user data that is added to the VM.")
+	cmd.MarkFlagsMutuallyExclusive(GenPasswordFlag, PasswordFileFlag)
+
+	cmd.Flags().StringVar(&c.cloudInitSourceType, CloudInitSourceTypeFlag, c.cloudInitSourceType, "Specify the cloud-init data source type. Allowed values are noCloud or configDrive.")
```
Comment (dominikholler) [inline pkg/virtctl/create/vm/vm.go:239]: Are we aware of any reason why the user might want to prefer configDrive over ncloud, or would it simplify the code if only nocloud would work?

Comment (0xFelix) [reply pkg/virtctl/create/vm/vm.go:239]: It was a TODO before to add support for config drives. IMO there is only slight overhead for supporting both so I'd keep it.

File: pkg/virtctl/create/vm/vm.go [hunk 2]
```diff
@@ -390,7 +428,16 @@ func (c *createVM) usage() string {
   {{ProgramName}} create vm --instancetype=my-instancetype --preference=my-preference --volume-pvc=my-pvc
 
   # Create a manifest for a VirtualMachine with a specified DataVolumeTemplate
-  {{ProgramName}} create vm --volume-import type:pvc,name:my-pvc,namespace:default,size:256Mi`
+  {{ProgramName}} create vm --volume-import type:pvc,name:my-pvc,namespace:default,size:256Mi
+
+  # Create a manifest for a VirtualMachine with a generated cloud-init config setting the user and adding an ssh authorized key
+  {{ProgramName}} create vm --user cloud-user --ssh-key"ssh-ed25519 AAAA...."
```
Comment (lyarwood) [inline pkg/virtctl/create/vm/vm.go:434]: `--ssh-key="ssh-ed25519 AAAA...."`

File: pkg/virtctl/create/vm/vm.go [hunk 3]
```diff
@@ -208,8 +229,21 @@ func NewCommand(clientConfig clientcmd.ClientConfig) *cobra.Command {
 		snapshot, params.Supported(dataVolumeSourceSnapshot{}),
 	))
 
+	cmd.Flags().StringVar(&c.user, UserFlag, c.user, "Specify the user in the cloud-init user data that is added to the VM.")
+	cmd.Flags().BoolVar(&c.genPassword, GenPasswordFlag, c.genPassword, "Specify if a password should be generated in the cloud-init user data that is added to the VM.")
+	cmd.Flags().StringVar(&c.passwordFile, PasswordFileFlag, c.passwordFile, "Specify a file to read the password from for the cloud-init user data that is added to the VM.")
+	cmd.Flags().StringSliceVar(&c.sshKeys, SSHKeyFlag, c.sshKeys, "Specify one or more SSH authorized keys in the cloud-init user data that is added to the VM.")
+	cmd.Flags().BoolVar(&c.gaManageSSH, GAManageSSHFlag, c.gaManageSSH, "Specify if the qemu-guest-agent should be able to manage SSH in the cloud-init user data that is added to the VM.")
```
Comment (lyarwood) [inline pkg/virtctl/create/vm/vm.go:236]: I don't think this is useful for now unless we dump secrets and wire up the actual passing of keys through the agent. I'd drop this for now tbh.

File: pkg/virtctl/create/vm/vm.go [hunk 4]
```diff
@@ -556,6 +603,119 @@ func (c *createVM) getInferFromVolume(vm *v1.VirtualMachine) (string, error) {
...
+	if c.genPassword {
+		password = rand.String(12)
+	}
+	if c.passwordFile != "" {
+		data, err := os.ReadFile(c.passwordFile)
+		if err != nil {
+			return "", params.FlagErr(PasswordFileFlag, "%w", err)
+		}
+		password = strings.TrimSpace(string(data))
+	}
```
Comment (lyarwood) [inline pkg/virtctl/create/vm/vm.go:700]: I'm honestly not sure about dumping raw passwords in the VM definitions like this.

Anyone with read access on the VMs is going to see this right?

File: pkg/virtctl/create/vm/vm.go (hunk 3 above)
Comment (0xFelix) [reply pkg/virtctl/create/vm/vm.go:236]: It is useful in combination with `virtctl credentials add-ssh-key`, which already makes use of the agent to sync authorized keys.

File: pkg/virtctl/create/vm/vm.go (hunk 4 above)
Comment (lyarwood) [reply pkg/virtctl/create/vm/vm.go:700]: Right, given our medium for storing and transmitting this config is a VirtualMachine object in k8s I really don't think we should directly enable this through `virtctl`.

File: pkg/virtctl/create/vm/vm.go (hunk 3 above)
Comment (lyarwood) [reply pkg/virtctl/create/vm/vm.go:236]: Ah, can you add that context to the help text?

File: pkg/virtctl/create/vm/vm.go [hunk 5]
```diff
@@ -556,6 +605,127 @@ func (c *createVM) getInferFromVolume(vm *v1.VirtualMachine) (string, error) {
...
+		return nil
+	}
+
+	const passwordWarningFmt = "WARNING: --%s: The password is stored in cleartext in the VM definition!\n"
+	if cmd.Flags().Changed(GenPasswordFlag) {
+		cmd.PrintErrf(passwordWarningFmt, GenPasswordFlag)
+	}
+	if cmd.Flags().Changed(PasswordFileFlag) {
+		cmd.PrintErrf(passwordWarningFmt, PasswordFileFlag)
+	}
```
Comment (0xFelix) [inline pkg/virtctl/create/vm/vm.go:626]: @lyarwood

File: pkg/virtctl/create/vm/vm.go (hunk 3 above)
Comment (0xFelix) [reply pkg/virtctl/create/vm/vm.go]: Done

File: pkg/virtctl/create/vm/vm.go (hunk 4 above)
Comment (lyarwood) [reply pkg/virtctl/create/vm/vm.go:710]: Yup that's a valid comprimise for now given it's already possible for users to do this outside of `virtctl`.

**Review by lyarwood — APPROVED** [review]
/lgtm

File: pkg/virtctl/create/vm/vm.go [hunk 6]
```diff
@@ -208,8 +232,21 @@ func NewCommand(clientConfig clientcmd.ClientConfig) *cobra.Command {
 		snapshot, params.Supported(dataVolumeSourceSnapshot{}),
 	))
 
+	cmd.Flags().StringVar(&c.user, UserFlag, c.user, "Specify the user in the cloud-init user data that is added to the VM.")
+	cmd.Flags().BoolVar(&c.genPassword, GenPasswordFlag, c.genPassword, "Specify if a password should be generated in the cloud-init user data that is added to the VM.")
```
[the remark this answers is not in the document]
Comment (0xFelix) [reply pkg/virtctl/create/vm/vm.go]: I dropped the `--gen-password` flag entirely.

**Comment by jcanocan** [comment]
Nice effort!
/lgtm

**Pull request merged**